File: decompile/ScriptExporter.py
import os

def read7BitEncodedInt(data_buffer, offset):
    result = 0
    offset_len = 0
    bits_read = 0
    while True:
        if bits_read >= 35:
            raise ValueError("Invalid 7-bit encoded integer.")
        byte = data_buffer[offset]
        offset += 1
        offset_len += 1
        result |= (byte & 0x7F) << bits_read
        bits_read += 7
        if (byte & 0x80) == 0:
            break
    return {
        "offset_len": offset_len,
        "content_length": result
    }

def extractFiles(script_file,callback):
    with open(script_file, 'rb') as f:
        data_buffer = f.read()

    offset = 0
    # 读取第一个 32 位整数（忽略）
    offset += 4
    # 读取第二个 32 位整数（忽略）
    offset += 4
    # 读取版本号
    version = int.from_bytes(data_buffer[offset:offset+4], byteorder='little')
    offset += 4
    # 读取文件数量
    count = int.from_bytes(data_buffer[offset:offset+4], byteorder='little')
    offset += 4

    for i in range(count):
        encode_info = read7BitEncodedInt(data_buffer, offset)
        offset += encode_info["offset_len"]
        name_length = encode_info["content_length"]
        name = data_buffer[offset:offset+name_length].decode('utf-8')
        offset += name_length

        content_length = int.from_bytes(data_buffer[offset:offset+4], byteorder='little')
        offset += 4

        content = data_buffer[offset:offset+content_length]
        offset += content_length

        save_file_name = os.path.join("export", name)
        directory = os.path.dirname(save_file_name)
        os.makedirs(directory, exist_ok=True)
        with open(save_file_name, 'wb') as out_file:
            out_file.write(content)
        callback(i,name,count)
    callback(count,"完成",count)

File: decompile/ScriptExporter.js
const fs = require('fs');
const path = require('path');

function read7BitEncodedInt(dataBuffer, offset)
{
    let result = 0;
    let offsetLen = 0;
    let bitsRead = 0;

    while (true)
    {
        if (bitsRead >= 35) throw new Error("Invalid 7-bit encoded integer.");

        const byte = dataBuffer.readInt8(offset++);
        offsetLen++;
        result |= (byte & 0x7F) << bitsRead;
        bitsRead += 7;

        if ((byte & 0x80) === 0) break;
    }

    return {
        offsetLen: offsetLen,
        contentLength: result,
    };
}

function extractFiles(scriptFile)
{
    const dataBuffer = fs.readFileSync(scriptFile);

    var offset = 0;
    dataBuffer.readInt32LE(offset);

    offset += 4;
    dataBuffer.readInt32LE(offset);
    offset += 4;
    var version = dataBuffer.readInt32LE(offset);
    offset += 4;
    var count = dataBuffer.readInt32LE(offset);
    offset += 4;
    for (var i = 0; i < count; i++)
    {
        var encodeInfo = read7BitEncodedInt(dataBuffer, offset);
        offset += encodeInfo.offsetLen;
        var nameLength = encodeInfo.contentLength;
        const name = dataBuffer.slice(offset, offset + nameLength).toString('utf-8');
        offset += nameLength;
        
        var contentLength = dataBuffer.readInt32LE(offset);
        offset += 4;
        
        var content = dataBuffer.slice(offset, offset + contentLength);
        offset += contentLength;

        var saveFileName = "export/" + name;
        const directory = path.dirname(saveFileName);
        fs.mkdirSync(directory, {recursive: true});
        fs.writeFileSync("export/" + name, content);
        console.log(`Extracted: ${name}`);
    }
}

const scriptFile = 'LWScripts.data';
extractFiles(scriptFile);
